components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use engine_traits::CacheRange;
+
+use crate::engine::SnapshotList;
+
+#[derive(Debug, Default)]
+pub struct RangeMeta {
+    range_snapshot_list: BTreeMap<CacheRange, SnapshotList>,
+    ranges_evcited: BTreeSet<CacheRange>,
+    ranges_unreadable: BTreeSet<CacheRange>,
+    safe_ts: u64,
+}
+
+impl RangeMeta {
+    pub(crate) fn range_snapshot_list(&self) -> &BTreeMap<CacheRange, SnapshotList> {
+        &self.range_snapshot_list
+    }
+
+    pub(crate) fn merge_meta(&mut self, mut other: RangeMeta) {
+        self.range_snapshot_list
+            .append(&mut other.range_snapshot_list);
+        self.ranges_evcited.append(&mut other.ranges_evcited);
+        self.ranges_unreadable.append(&mut other.ranges_unreadable);
+
+        // merge safe_ts to the min of them if not 0
+        if self.safe_ts != 0 && other.safe_ts != 0 {
+            self.safe_ts = u64::min(self.safe_ts, other.safe_ts);
+        } else if other.safe_ts != 0 {
+            self.safe_ts = other.safe_ts;
+        }
+    }
+
+    pub(crate) fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        if set_readable {
+            assert!(self.ranges_unreadable.remove(range));
+        } else {
+            self.ranges_unreadable.insert(range.clone());
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RangeManager {
+    // the range reflects the range of data that is accessable.
+    ranges: BTreeMap<CacheRange, RangeMeta>,
+}
+
+impl RangeManager {
```

Comment:
Just to make sure we're on the same page, I would use `overlaps_with_range` to find the right region cache in the write path?

Reply:
should it be contains, not overlaps_with_range.